Description: Cleanup stale device mapper mounts on start
  On shutdown, docker sometimes leaves device mapper
  mounts for container mounted, causing the containers
  impacted to be un-startable post-restart.
  .
  Cleanup any stale device mapper mounts on startup; this
  is really just a workaround for an open issue upstream.
Author: James Page <james.page@ubuntu.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1404300
Bug: https://github.com/docker/docker/issues/5684
Forwarded: no

To reproduce, just install docker.io from trusty archive (0.9.1 or 1.0.1), start a
container doing just a echo hello, every seconds, then upgrade. The container cannot
start again.

diff --git a/contrib/init/sysvinit-debian/docker b/contrib/init/sysvinit-debian/docker
index 0a5ffc8..d1a1158 100755
--- a/contrib/init/sysvinit-debian/docker
+++ b/contrib/init/sysvinit-debian/docker
@@ -59,12 +59,21 @@ fail_unless_root() {
 	fi
 }
 
+devicemapper_umount() {
+	# Cleanup any stale mounts left from previous shutdown
+	# see https://bugs.launchpad.net/ubuntu/+source/docker.io/+bug/1404300
+	grep "mapper/docker" /proc/mounts | awk '{ print $2 }' | \
+		xargs -r umount || true
+}
+
 case "$1" in
 	start)
 		check_init
 		
 		fail_unless_root
 
+		devicemapper_umount
+
 		touch "$DOCKER_LOGFILE"
 		chgrp docker "$DOCKER_LOGFILE"
 
diff --git a/contrib/init/upstart/docker.conf b/contrib/init/upstart/docker.conf
index 4414e4b..e8f718a 100644
--- a/contrib/init/upstart/docker.conf
+++ b/contrib/init/upstart/docker.conf
@@ -7,6 +7,13 @@ limit nproc 524288 1048576
 
 respawn
 
+pre-start script
+	# Cleanup any stale mounts left from previous shutdown
+	# see https://bugs.launchpad.net/ubuntu/+source/docker.io/+bug/1404300
+	grep "mapper/docker" /proc/mounts | awk '{ print $2 }' | \
+		xargs -r umount || true
+end script
+
 kill timeout 20
 
 script
